h2/src/main/org/h2/expression/SimpleCase.java

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright 2004-2020 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (https://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.expression;
+
+import org.h2.engine.Session;
+import org.h2.table.ColumnResolver;
+import org.h2.table.TableFilter;
+import org.h2.value.TypeInfo;
+import org.h2.value.Value;
+import org.h2.value.ValueNull;
+
+/**
+ * A simple case.
+ */
+public class SimpleCase extends Expression {
+
+    public static abstract class SimpleWhen {
+
+        Expression result;
+
+        SimpleWhen next;
+
+        SimpleWhen(Expression result) {
+            this.result = result;
+        }
+
+        public void addWhen(SimpleWhen next) {
+            this.next = next;
+        }
+
+    }
+
+    public static final class SimpleWhen1 extends SimpleWhen {
+
+        Expression operand;
+
+        public SimpleWhen1(Expression operand, Expression result) {
+            super(result);
+            this.operand = operand;
+        }
+
+    }
+
+    public static final class SimpleWhenN extends SimpleWhen {
+
+        Expression[] operands;
+
+        Expression result;
+
+        public SimpleWhenN(Expression[] operands, Expression result) {
+            super(result);
+            this.operands = operands;
+        }
+
+    }
```

Comment:
this SimpleWhen1 vs SimpleWhenN thing seems like a lot of code for not much benefit. Why not just have SimpleWhenN ?

Reply:
I'll merge them, you're right, specialized implementation for typical case with only one operand doesn't improve performance due to all these branches.